accepted/system-range/system-range.md

```diff
@@ -0,0 +1,246 @@
+* Creating overloads on existing types that will accept `Range` (that doesn't
+  mean we shouldn't be doing it, but that it will be covered by a separate
+  design proposal).
+* Supporting `Index` and `Range` on multi-dimensional arrays and non-zero based
+  arrays (e.g. instances of `System.Array`).
+    - This work needs more thought and is currently scoped out.
+
+## Design
+
+### Proposed API
+
+```C#
+namespace System
+{
+    public readonly struct Index
+    {
+        private readonly int _value;
+
+        public int Value => _value < 0 ? ~_value : _value;
+        public bool FromEnd => _value < 0;
+
+        public Index(int value, bool fromEnd = false)
+        {
+            if (value < 0)
+                throw new ArgumentException("Index must not be negative.", nameof(value));
+
+            _value = fromEnd ? ~value : value;
+        }
+
+        public static implicit operator Index(int value) => new Index(value);
+    }
+
+    public readonly struct Range
+    {
+        public Index Start { get; }
+        public Index End { get; }
+
+        private Range(Index start, Index end)
+        {
+            Start = start;
+            End = end;
+        }
+
+        public static Range Create(Index start, Index end) => new Range(start, end);
+        public static Range FromStart(Index start) => new Range(start, new Index(0, fromEnd: true));
+        public static Range ToEnd(Index end) => new Range(new Index(0, fromEnd: false), end);
+        public static Range All() => new Range(new Index(0, fromEnd: false), new Index(0, fromEnd: true));
+    }
+```
+
+### Indexing with SZ-arrays
+
+SZ-arrays (i.e. `T[]` in C#) do not have an API that represents the indexer.
+Instead, they are specialized as IL opcodes to load and store elements in such
+an array. The indexer is part of the language semantics and thus doesn't have an
+API representation.
+
+For indexing using `Index` the compiler doesn't have to do much, thus the
+handling is done directly by the compiler, i.e. this code:
+
+```C#
+T[] array = GetArray();
+Index index = GetIndex();
+T element = array[index];
+```
+
+will be compiled down to the following code:
+
+```C#
+T[] array = GetArray();
+Index index = GetIndex();
+int $index = index.FromEnd
+                ? array.Length - index.Value
+                : index.Value;
+T element = array[$index];
+```
+
+For `Range` the code-gen would be a more a bit more involved, thus we propose
+to give the compiler a `RuntimeHelper` API to call, so this code:
+
+```C#
+T[] array = GetArray();
+Range range = GetRange();
+T[] elements = array[range];
+```
+
+will be compiled down to:
+
+```C#
+T[] array = GetArray();
+Range range = GetRange();
+T[] elements = RuntimeHelpers.GetArrayRange<T>(array, range);
+```
+
+We propose the following method:
+
+```C#
+namespace System.Runtime.CompilerServices
+{
+    public sealed class RuntimeHelpers
+    {
+        public static T[] GetArrayRange<T>(T[] array, Range range);
```

Comment:
Does this return a _new_ array with the data _copied_ out of `array`? If so, that seems wrong.

Instead, I would assume this should return `Span<T>`.

Reply:
> Does this return a new array with the data copied out of array? If so, that seems wrong.

If it's being used by the compiler to generate the code for:
```
T[] elements = array[range];
```
as above, then returning a span doesn't help.

Reply:
ps This is based on the idea that a) we should have range-based indexers on array and string, and b) that all range-based indexers should return the same type as the one they're on.  I argued that indexers should be non-allocating/non-copying, and so if we want to stick with (b), then we shouldn't have these indexers on array and string.  I lost.

Reply:
It's frustrating that we spend so much effort on non-allocating scenarios, only to bring them back front and center in new scenarios.

Reply:
Yes, yes it is.  I'll assume your thumbs down on my comment isn't about my opinion but about the fact that I lost.

Reply:
In my mind, `string` is a different scenario than `array` and they shouldn't be grouped together.  (strings have semantics - like equality - that arrays don't, so you can't say "string" is just an array of characters. If it was, we would have just used `char[]`. Also, arrays are mutable and strings aren't.)

With string, you can call `Substring`, and that is well known to be allocating a new string. Also, why do we have _both_ the indexer `[Range]` and `Substring` on `string`? They do the same thing right? Why would I choose one over the other?

So instead, `string[Range]` could return `ReadOnlyMemory|Span<char>` and `string.Substring(Range)` return `string`. The difference is clear.

Then `array[Range]` could return `Span<T>` (or `ArraySegment<T>`).

Honestly, I don't see a need for the `array.Subarray(Range)` scenario. If you really want that - allocate a new array, and call `Array.Copy`.

> about the fact that I lost.

Correct.

Reply:
>  I argued that indexers should be non-allocating/non-copying

:+1:

The allocating array indexers look like a  major performance trap. I do not think we should have them.

Reply:
> The allocating array indexers look like a major performance trap.

Yes, especially since if you write code like:
```C#
string s = "hello, world";
ReadOnlySpan<char> span = s[1..3];
```
it's far from obvious that allocates and is different from using AsSpan(…).

Reply:
This was discussed in the C# LDM last week. I presented my opinion which is as follows:

1. Arrays and strings are vastly more useful to the majority of our customers than span-based APIs.
2. The rules for the range indexer are much simpler if all types follow the same pattern, namely that they return the same type as the container.
3. Performance, while certainly a major focus for us, isn't the most important goal for the majority of code that is written. C# has plenty of features that implicitly allocate. While we often see those as problematic for performance critical code, they are also a major reason why C# is loved as a productive language.

It seemed the overwhelming majority of people in the room shared that opinion. That being said, I don't believe the issue that Stephen pointed out was considered at the time. But this strikes me as yet another instance of an analyzer that we should offer. The offered fix should be `ReadOnlySpan<char> span = s.AsSpan(1..3);`

Reply:
The other argument that Mads brought is that we should think of indexers with ranges as a different thing, if we believe this deserves compact syntax. In the end, one of the biggest customers of range will be `Tensor<T>`. Slicing that guy won't be an O(1) allocation-free operation either.

If we truly believe that index-syntax must mean allocation-free and O(1) access, just removing the range-based indexer from array alone isn't the right fix, we'd also need to remove it from string. However, I suspect that this means the compacter syntax benefits much fewer users. The difference between `s.Substring(start, length)` and `s.Substring(start..end)` isn't nearly as significant as `s[start..end]`.

Reply:
> C# has plenty of features that implicitly allocate.

Yes, that's correct. However, I believe that it is important that the rought performance characteristics are predictable by looking at the code without thinking about it too hard. It is part of the balance that makes C# good productive language.

For example, if you see Linq, you know that you are on the slow plan. If you see `async`, you know that you are paying some overhead, but it is not terrible. The allocating/copying indexers fail on this.

Reply:
> In the end, one of the biggest customers of range will be Tensor<T>. Slicing that guy won't be an O(1) allocation-free operation either.

I implemented `Tensor<T>.Slice` to not copy data. You can just change the strides you are using, and now you can project into the same data without making a copy.

See https://github.com/dotnet/corefxlab/blob/38fb993f61467f24fb845323f25e807c4a618827/src/System.Numerics.Tensors/System/Numerics/Tensors/Tensor.cs#L1272

[numpy does the same thing](https://scipy-cookbook.readthedocs.io/items/ViewsVsCopies.html).

From the numpy link above:

> Technically, that means that the data of both objects is shared. You can create views by selecting a slice of the original array

So it is definitely not the case that slicing a Tensor will copy the data out.
__________

> The rules for the range indexer are much simpler if all types follow the same pattern, namely that they return the same type as the container.

That does not hold for all cases. It definitely doesn't hold for the `DenseTensor` case, where you project into the same data.

Imagine you are processing a video (or 3D image) using Tensor, and you wanted to slice out a single image in order to do some processing on it. Forcing a copy of the data for every frame in the video is not going to produce great performance.

Reply:
> I implemented `Tensor<T>.Slice` to not copy data. You can just change the strides you are using, and now you can project into the same data without making a copy.

I didn't say copy, I said not cheap (compared to returning span or an element). Returning views that can mutate the original data have their own problems and can surprise people too. But that's a different problem we should discuss when reviewing tensors.

> > The rules for the range indexer are much simpler if all types follow the same pattern, namely that they return the same type as the container.
> 
> That does not hold for all cases. It definitely doesn't hold for the `DenseTensor` case, where you project into the same data.

That depends on how it's exposed. Looking at the thing you linked, it looks like the `Tensor<T>` class has a a virtual `Slice` method that is typed to return a `Tensor<T>`. I'd say that's close enough to what I said.

Reply:
> I didn't say copy, I said not cheap 

You said `won't be an O(1) allocation-free operation`. The current implementation of `Tensor<T>.Slice` is `O(1)`, it is not allocation-free however because Tensor is a class.

> That depends on how it's exposed. Looking at the thing you linked, it looks like the Tensor<T> class has a a virtual Slice method that is typed to return a Tensor<T>. I'd say that's close enough to what I said

When a user explicitly has a `DenseTensor<float>` and they slice it, they won't be getting back a `DenseTensor<float>`. Instead they will get a view into the current `DenseTensor`, which is the base type `Tensor<float>`. So here it isn't the same type as the container.

Reply:
Alright, for now I'm going to resolve this as the current spec lists the POR as agreed between FXDC/LDM. We can re-evaluate this based on customer feedback in Preview 2.

Reply:
Would it make more sense to not add questionable indexers first, and only add them later if the customer feedback tells us that they are really needed?